keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md

```diff
@@ -0,0 +1,163 @@
+
+CSI topology labels could be transformed to kubernetes labels within the permitted topology label namespace:
+
+* CSI topology labels that did not have a prefix would have a prefix of `topology.kubernetes.io/` prepended
+* CSI topology labels with a prefix not ending in `topology.kubernetes.io` would have that string added to their prefix
+
+Component behavior:
+
+* The kubelet would transform CSI topology to kubernetes labels before adding them to the `Node` API object.
+* The CSI provisioner would transform CSI topology labels to kubernetes labels when [retrieving topology info from a given node](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305).
+* The CSI provisioner would transform kubernetes labels into CSI topology labels when [building required topology from a storage class AllowedTopologies](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L75-L91).
+* The associated storage class would continue to specify topology in terms of kubernetes label selectors.
+
+Example 1:
+
+* CSI driver returns topology labels `{"acme.com/zone":"A"}`
+* The kubelet records `["acme.com/zone"]` in CSINodeInfo.CSIDriverInfo.TopologyKeys
+* The kubelet transforms the topology info to a `{"acme.com.topology.kubernetes.io/zone":"A"}` label and adds it to the Node object
+* The storage class lists allowed topology selectors of `{"key":"acme.com.topology.kubernetes.io/zone","values":["A"]}`
+* The CSI provisioner transforms the topology key of `"acme.com/zone"` to a label name of `"acme.com.topology.kubernetes.io/zone"` in [getTopologyFromNode](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305)
+* The CSI provisioner transforms the storage class selector of `{"key":"acme.com.topology.kubernetes.io/zone","values":["A"]}` to a topology requirement of `{"acme.com/zone":"A"}` when creating the volume
+
+Example 2:
+
+* CSI driver returns topology labels `{"zone":"A"}`
+* The kubelet records `["zone"]` in CSINodeInfo.CSIDriverInfo.TopologyKeys
+* The kubelet transforms the topology info to a `{"topology.kubernetes.io/zone":"A"}` label and adds it to the Node object
+* The storage class lists allowed topology selectors of `{"key":"topology.kubernetes.io/zone","values":["A"]}`
+* The CSI provisioner transforms the topology key of `"zone"` to a label name of `"topology.kubernetes.io/zone"` in [getTopologyFromNode](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305)
+
+## Implementation Timeline
+
+v1.13:
+
+* Kubelet deprecates setting labels other than the ones described above via `--node-labels`, and warns when invoked with labels outside that set
+* NodeRestriction admission prevents kubelets from setting labels other than the ones described above on Node *update* (not on Node create)
```

Comment:
And this validation only runs if the labels were changed?

Reply:
correct.

Reply:
So we are saying that the comprised node, can only try to update the labels for its own node, not add new labels ? What if the comprised node can register as a new node ?

Reply:
> So we are saying that the comprised node, can only try to update the labels for its own node, not add new labels ?

Preventing labels outside a known set can be done on update immediately because kubelets do not update arbitrary labels on their node objects today.

> What if the comprised node can register as a new node?

Letting nodes register as arbitrary new/different nodes is not recommended.

Reply:
i dont fully understand this . If a node gets compromised today, it can modify the topology label of that node and so attract arbitrary workloads which were meant for other nodes. How does this feature prevent that. The node is still allowed to update the topology label(one of the allowed ones). What we want is the node only allows a fix set of labels and they can never be updated for a given node. 
 
Also my other doubt (mainly because i dont understand node authorizer properly ) is can a  compromised node ever register as a new node ?

Reply:
> If a node gets compromised today, it can modify the topology label of that node and so attract arbitrary workloads which were meant for other nodes. How does this feature prevent that.

The preferred way of labeling node objects is from a central authoritative source. Bounding the labels kubelets can self-apply means that centrally-set labels outside those prefixes can be trusted to not be settable by compromised nodes.

For example, labels that isolate nodes by tenant, by security policy, by trust level, by workload sensitivity, etc, are all things that are orthogonal to topology, and should be set top-down and be unchangeable by kubelets.

> can a compromised node ever register as a new node 

The node authorizer and noderestriction admission plugin limit kubelets to registering a node object that matches their identity.